src/printer.js

```diff
@@ -338,12 +338,16 @@ function genericPrintNoParens(path, options, print) {
 
     parts.push(" =>");
 
-    return conditionalGroup([
-      concat([ concat(parts), " ", path.call(print, "body") ]),
-      concat([ concat(parts),
-               indent(options.tabWidth,
-                      concat([line, path.call(print, "body")]))])
-    ])
+    const groups = [
+      concat([ concat(parts), " ", path.call(print, "body") ])
+    ];
+    if (n.body.type !== "JSXElement") {
+      groups.push(concat([
+        concat(parts),
+        indent(options.tabWidth, concat([ line, path.call(print, "body") ])),
+      ]));
+    }
+    return conditionalGroup(groups);
```

Comment:
I would prefer avoiding `conditionalGroup` in this special case which makes this code path a little clearer:

```js
const body = path.call(print, "body");
const collapsed = concat([ concat(parts), " ", body ]);

if (n.body.type === "JSXElement") {
  return group(collapsed);
}

return conditionalGroup([
  collapsed,
  groups.push(concat([
    concat(parts),
    indent(options.tabWidth, concat([ line, body ])),
  ]));
])
```

Also note that I lifted `path.call(print, "body")` out which is very important. I wrote that code so it's my fault, but when using `conditionalGroup` you should never print the same node twice because it makes the intermediate representation explode in size (we want to share sub-trees).